PowerEditor/src/Parameters.cpp

```diff
@@ -1939,9 +1943,442 @@ bool NppParameters::getUserStylersFromXmlTree()
+					{
+						std::wstring wsToken;
+						std::wistringstream wstrm(wsTextModel);
+						while (wstrm >> wsToken)
+						{
+							if (!mapUserWords.contains(wsToken))
+							{
+								vwsUserWords.push_back(wsToken);
+								mapUserWords[wsToken] = true;
+								++nWordsAdded;
+							}
+						}
+					}
+
+					// if there were any words added to the list, need to update the element contents
+					if (nWordsAdded)
+					{
+						// sort the words in standard case-sensitive alphabetical order
+						std::sort(vwsUserWords.begin(), vwsUserWords.end());
+
+						// convert that list into space-separated string, with at most 8000 characters per line
+						size_t lineLength = 0, maxLineLength = 8000;
+						bool first = true;
+						std::wstring wsOutputWords(L"");
+						for (auto wsWord : vwsUserWords)
+						{
+							if (!first)
+							{
+								// space between words
+								wsOutputWords += L" ";
+								lineLength += 1;
+							}
+							first = false;
+
+							if (lineLength + wsWord.length() >= maxLineLength)
+							{
+								// start next line
+								lineLength = 0;
+								wsOutputWords += L"\n                ";
+							}
+
+							// add this word to the output string
+							wsOutputWords += wsWord;
+							lineLength += wsWord.length();
+						}
+
+						// and update the XML's value
+						pChild->SetValue(wsOutputWords);
+					}
+				}
+				else
+				{
+					// if this Keywords element doesn't exist in user list, need to clone it from model to the right parent language in the user list
+					TiXmlNode* p_clone = keywordsFromModel->Clone();
+					mapUserLanguages[modelLanguageName]->LinkEndChild(p_clone);
+				}
+			}
+
+			// Also, since <Language name="..." ...> can have other attributes, need to check to make sure that
+			//		the user langs copy of language isn't missing any of the attributes from the model
+			TiXmlElement* thisLanguageFromUser = mapUserLanguages[modelLanguageName];
+			for (TiXmlAttribute* attrModel = langFromModel->FirstAttribute();
+				attrModel;
+				attrModel = attrModel->Next())
+			{
+				// if attribute not in user, need to add it (but leave it alone if the user-langs has it, but is just an empty string, because that's intentionally blank)
+				const wchar_t* pwcUserValue = thisLanguageFromUser->Attribute(attrModel->Name());
+				if (!pwcUserValue)
+					thisLanguageFromUser->SetAttribute(attrModel->Name(), attrModel->Value());
+			}
+		}
+		else
+		{
+			// otherwise, since Language doesn't exist in User Languages, need to duplicate/clone from model to user-langs structure
+			TiXmlNode* p_clone = langFromModel->Clone();
+			langsTopUser->LinkEndChild(p_clone);
+		}
+	}
+
+	// check the user-langs document for changes
+	sUserText = "";
+	_pXmlDoc->Print(sUserText);
+	std::string md5digest_user_text_after = md5.digestString(sUserText.c_str());
+	if (md5digest_user_text_before != md5digest_user_text_after)
+		_pXmlDoc->SaveFile();
+
+	delete pXmlModel;
+	return;
+}
+

```

Comment:
The both functions
`void NppParameters::updateKeyWordsFromModelXml(TiXmlNode* rootUser)`
&
`void NppParameters::updateUserStylersFromModelXml(TiXmlNode* rootUser)`
share the same procedure:
1. Load the model XML file
2. Computer its MD5 & compare it
3. If necessary, update the destination file
4. Save & clean up (de-allocation)

Only the point 3 is not in common for these 2 functions.

How about sharing the same function `updateFromModelXml(TiXmlNode* rootUser, ConfXml whichConf)`, by using `enum ConfXml {lang, styles};`.
And in this function, according whichConf, you call `updateLangXML(...)` or `updateStylesXML(...)` ?
In this way, the code is more readable, and we could add more update XML (nativeLang.xml eventually) into `updateFromModelXml`in the future.

Reply:
Good idea.  My newest commit implements this.